### PR TITLE
[feat] refresh token

### LINKED ISSUE #220: [FEAT] Refresh expired token
**💡 I would really like to solve or include**   
Refresh expired token and black-list the expired one.

**👶 How would a user describe this?**   
I would like to stay logged-in until I choose to log-out. 

**🚀 I'm ready for take off**  
Before submitting, please mark if you:  
- [x] Checked that this feature doesn't already exists  
- [x] Checked that a feature request doesn't already exists  
- [x] Went through the user flow, and understand the impact  
- [x] Made sure the request shows why it is important to users but doesn't exaggerate the value

### PR DESCRIPTION
**Pull request by nannadao**
Implement refresh token in private route.
To test out, remember to add `JWT_REFRESH_SECRET` variable to your .env
Currently I set the access token to expire in 5 mins so you have shorter waiting time. Later on I can extend it to 1hr.

Update 1: It is possible to refresh page and get the refreshed token now. 

Update 2: A new access token is now generated from isAuthenticated middleware. Also, if refreshToken expired, I removed the user from cookies (same with logout function). Please inform me anything you'd like to improve!

Close #220

### CODE REVIEW

File: src/helpers/generateToken.ts [hunk 1]
```diff
@@ -1,16 +1,25 @@
 import jwt from 'jsonwebtoken'
 
-import { JWT_SECRET } from '../util/secrets'
-
-const generateToken = (id: string) => {
+import { JWT_SECRET, JWT_REFRESH_SECRET } from '../util/secrets'
+//Generate access Token
```
Comment (Rostislav013) [inline src/helpers/generateToken.ts:4]: Shall it be a space between there?

**Review by Rostislav013 — APPROVED** [review]

**Review by MikeVedsted — APPROVED** [review]
I think you accomplish what you set out to do with the scope of the ticket.
As improvements, you could maybe move the getUserInfo from ProtectedRoute to a hook since you are reusing the same logic in ProfilePage. 
Additionally, I  don't think we need to have two additional routes to handle this logic but rather that it could be implemented in the isAuthenticated middleware. 

I am having one issue with your request. When accessing a profile page the first time, all account values are undefined. I have to refresh first to get the actual values, regardless of my token status. You might want to check your async functions.

Your PR highlights a new issue. When the refresh token is expired, the user can currently still browse the site (just without the protected data). The user should however be taken to log in or get started, once no valid token exists.

File: client/src/pages/ProfilePage/index.tsx [hunk 2]
```diff
@@ -75,38 +76,32 @@ const ProfilePage = () => {
...
           <h4 className='profile__details--header'> Full Name: </h4>
           <p className='profile__details--text'>
-            {' '}
-            {`${userInfo.first_name} ${userInfo.last_name}`}{' '}
+            {`${userInfo.first_name} ${userInfo.last_name}`}
           </p>
         </div>
-
         {cookies.user.user_id === userId && (
           <div className='profile__details'>
```
Comment (Rostislav013) [inline client/src/pages/ProfilePage/index.tsx:83]: Instead `cookies.user.user_id` can use `currentUser` cuz it is declared alr.

File: client/src/pages/ProfilePage/index.tsx [hunk 3]
```diff
@@ -75,38 +76,32 @@ const ProfilePage = () => {
...
           </p>
         </div>
-
         <div className='profile__details'>
           <h4 className='profile__details--header'> Intro: </h4>
           <p className='profile__details--text'> {userInfo.profile_text} </p>
         </div>
-
         {cookies.user.user_id === userId && (
           <div className='profile__details'>
```
Comment (Rostislav013) [inline client/src/pages/ProfilePage/index.tsx:99]: currentUser

**Review by Rostislav013 — APPROVED** [review]

File: src/helpers/generateToken.ts (hunk 1 above)
Comment (MikeVedsted) [reply src/helpers/generateToken.ts:4]: yes

File: src/helpers/generateToken.ts [hunk 4]
```diff
@@ -1,16 +1,25 @@
...
+export const generateAccessToken = (id: string) => {
   return jwt.sign(
     {
       iss: 'JoinMe',
       sub: id
     },
     JWT_SECRET,
-    { expiresIn: '1h' }
+    { expiresIn: 60 * 60 }
```
Comment (MikeVedsted) [inline src/helpers/generateToken.ts:12]: please lower to 15 minutes

File: src/controllers/user.ts [hunk 5]
```diff
@@ -13,6 +13,18 @@ export const googleLogin = async (req: Request, res: Response, next: NextFunctio
...
+    if (!req.user) {
+      throw Error
+    }
+    const { user_id } = req.user
+    return res.json(await UserService.findUserById(user_id))
+  } catch (error) {
+    next(new BadRequestError('Unexpected error', error))
+  }
+}
+
```
Comment (MikeVedsted) [inline src/controllers/user.ts:27]: You don't actually need to do anything else than trigger the middleware, right? So you don't actually need this function.

**Review by MikeVedsted — APPROVED** [review]

File: src/controllers/user.ts (hunk 5 above)
Comment (nannadao) [reply src/controllers/user.ts:27]: This function is also to return user's info and update cookies

File: client/src/hooks/useHostedEvents.tsx [hunk 6]
```diff
@@ -7,8 +7,8 @@ const useHostedEvents = (userId: UserId) => {
   const [hostedEvents, setHostedEvents] = useState([])
 
   useEffect(() => {
-    fetchHostedEvents()
-  }, [])
+    userId && fetchHostedEvents()
+  }, [userId])
```
[the remark this answers is not in the document]
Comment (nannadao) [reply client/src/hooks/useHostedEvents.tsx:12]: it will unfortunately trigger a loop

File: src/controllers/user.ts (hunk 5 above)
Comment (MikeVedsted) [reply src/controllers/user.ts:27]: Why not use the existing function then? You have just copied findUserById and given it a different name

Comment (nannadao) [reply src/controllers/user.ts:27]: the difference is the findUserById take userId from params, here I take userId from req.user. But after having the userId, the service is the same, hence they only have different controller

**Pull request merged**